lib/model/model.go

```diff
@@ -250,7 +253,18 @@ func (m *model) onServe() {
 			continue
 		}
 		m.newFolder(folderCfg)
-	}
+		// Forget pending folders that are now added
+		m.db.RemovePendingFolder(folderCfg.ID, folderCfg.DeviceIDs())
+	}
+	// Unique set of device IDs for which we'd like to keep pending folder entries
+	keepPendingFoldersFor := make(map[protocol.DeviceID]bool, len(m.cfg.Devices()))
+	for deviceID := range m.cfg.Devices() {
+		// Forget pending devices that are now added
+		m.db.RemovePendingDevice(deviceID)
+		keepPendingFoldersFor[deviceID] = true
+	}
+	// Clean pending folder entries for devices no longer known
+	m.db.CleanPendingFolders(keepPendingFoldersFor)
```

Comment:
Is `CommitConfiguration()` called when the configuration file was edited while Syncthing had not been running?  That's the case I had in mind, to do a clean-up pass right on startup, just like what previously happened in `config.prepare()` / `clean()`.

At that point, there is no "old" vs. "new" configuration, so I need the function `CleanPendingFolders()` which only knows what is no longer pending but properly added / ignored.

Reply:
You're right it's not - this is good then.

Different point: Shouldn't you also consider ignored devices here? Maybe there's an option to move some of this into a helper function used both here and commitconf.

Reply:
Yes, ignored devices should be processed here as well, already had a comment for that locally.

To get an overview for myself, I summarized all possible configuration changes and the related actions regarding pending device / folder entries: 
![pending-cleanup](https://user-images.githubusercontent.com/6996887/79055438-5fcfc780-7c4d-11ea-8791-4a4ff8f5c3ea.png)

There are three special cases where `onServe()` and `CommitConfiguration()` may behave differently:
1. Devices no longer in the configuration cannot be enumerated in `onServe()`, so it's unclear whether they were previously ignored or properly known.  In the latter case (device removed) there should not be a pending device entry at all, so it's safe not trying to delete it.  In contrast, `CommitConfiguration()` can distinguish the two cases and could clean up after a removed device to avoid a new pending notification (which is expected and desired for unignored devices though, so keep).

2. Entries for pending folders always have a device ID associated.  If that is not among the configured device IDs because the device was ignored or removed, these pending folders should be cleaned up.  `onServe()` has no "old" config to compare, so must iterate over all pending folder entries and filter to keep only those for known devices.

3. If a folder configuration is completely removed (not just a device from it), we assume that it is no longer of interest and delete all its pending folder entries.  If a known device still offers the folder to us, the pending entry will come back on next connection.  However again, `onServe()` does not know about removed folders, so the intention is not clear and no entries should be deleted.

Hopefully this table is comprehensible, otherwise please ask.  I think it covers all cases, so I'll rework and check my implementation against it when we agree on the appropriate cleanup actions.

Reply:
2. keep for known *and pending* devices.

3. There shouldn't be any pending folder entries for a known folder, they should be removed on adding (and no new ones added), right?

Otherwise sounds good.

Reply:
2. I'll have to check what the current behavior was, but as long as a device is not added locally, the connection is closed after the Hello message without even receiving a ClusterConfig that might create pending folder entries. So I assumed that pending folders should only exist for known devices.

3. There can be pending folder entries for a locally available folder which is not shared to all devices that offer it to us. In that case they are only offered as a "share with" notification instead of an "add folder" notification.

Reply:
Right, thanks for the correction - <strike>Otherwise</strike> All sounds good then :)